### PR TITLE
rg-63: Forgot password

### PR DESCRIPTION
**Pull request by vlad-blip**
Bug LP-6 was fixed there as well.

### CODE REVIEW

**Comment by vlad-blip** [comment]
[Forgot password](https://trello.com/c/YJgGLAWx/63-forgot-password)

**Comment by vlad-blip** [comment]
[The "Forgot password" button is not clickable on the "Log In" page.](https://trello.com/c/cio1R7OZ/108-the-forgot-password-button-is-not-clickable-on-the-log-in-page)

File: backend/src/bundles/auth/auth.service.ts
Comment (Vitaliy-Kalenichenko) [inline backend/src/bundles/auth/auth.service.ts:1]: Looks like we already have functions to generate/issue tokens.

File: frontend/src/bundles/auth/components/password-recovery/constants/default-reset-token-payload.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { type UserVerifyResetTokenRequestDto } from 'shared/build/index.js';
+
+const DEFAULT_RESET_TOKEN_PAYLOAD: Omit<
+    UserVerifyResetTokenRequestDto,
+    'email'
+> = {
+    resetToken: '',
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/auth/components/password-recovery/constants/default-reset-token-payload.ts:7]:
```suggestion
    resetPasswordToken: '',
```

**Review by Vitaliy-Kalenichenko — CHANGES_REQUESTED** [review]
Check class names for compliance with the BEM methodology

File: backend/src/bundles/auth/auth.service.ts
Comment (vlad-blip) [reply backend/src/bundles/auth/auth.service.ts:1]: The existing token generators are hardcoded to authorization and refresh token config values. They have expire time of 24 hours or 30 days, whereas I need 10 minutes for reset password token.

File: backend/src/common/mail-service/mail-service.package.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ class MailService implements IMailService {
         const transportOptions: Options = {
             host: process.env.SMTP_HOST,
             port: Number(process.env.SMTP_PORT),
-            secure: true,
+            secure: false,
```
Comment (MaxPoplavskiy) [inline backend/src/common/mail-service/mail-service.package.ts:14]: Our coaches wanted it to be true, as I remember

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

File: backend/src/common/mail-service/mail-service.package.ts (hunk 2 above)
Comment (vlad-blip) [reply backend/src/common/mail-service/mail-service.package.ts:14]: You are correct, I didn't mean to push it.

**Review by MaxPoplavskiy — APPROVED** [review]

File: backend/src/common/rate-limit/constants/reset-password.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+const RESET_PASSWORD = {
+    MAX_ATTEMPTS: 3,
+    TIMEOUT: 600_000,
+} as const;
```
Comment (ValentinaTenus) [inline backend/src/common/rate-limit/constants/reset-password.ts:4]: consider using enums

File: shared/src/bundles/users/validation-schemas/email.validation-schema.ts
Comment (ValentinaTenus) [inline shared/src/bundles/users/validation-schemas/email.validation-schema.ts:1]: shouldn't it include requirements from https://docs.google.com/spreadsheets/d/1uGz_PZiFPNCPIZqIX5L_ZBdgje4URtWwD-EauYSG2fw/edit#gid=0

File: shared/src/bundles/users/validation-schemas/restore-password.validation-schema.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import joi from 'joi';
+
+import { type UserResetPassword } from '../types/user-reset-password.type.js';
+import { UserValidationMessage } from '../users.js';
+
+const passwordValidationSchema = joi.object<UserResetPassword, true>({
+    password: joi.string().regex(/^\S*$/).min(8).max(64).messages({
```
Comment (ValentinaTenus) [inline shared/src/bundles/users/validation-schemas/restore-password.validation-schema.ts:7]:
```suggestion
     password: joi
        .string()
        .regex(
            /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!"#$%&'()*+,./:;<=>?@[\\\]^_`{|}~])(?=\S{8,64}$).*$/,
        )
        .min(UserValidationRule.PASSWORD_MIN_LENGTH)
        .max(UserValidationRule.PASSWORD_MAX_LENGTH)
        .required()
        .messages({
```

File: shared/src/bundles/users/validation-schemas/email.validation-schema.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        .email({
+            tlds: {
+                allow: false,
+            },
+        })
+        .required()
+        .messages({
+            'string.email': UserValidationMessage.EMAIL_WRONG,
+            'string.empty': UserValidationMessage.EMAIL_REQUIRE,
+        }),
```
Comment (ValentinaTenus) [inline shared/src/bundles/users/validation-schemas/email.validation-schema.ts:21]:
```suggestion
        email: joi
        .string()
        .email({
            tlds: {
                allow: false,
            },
        })
        .custom((value, helpers) => {
            const localPart = value.split('@')[0];
            const domain = value.split('@')[1];

            if (localPart.length <= 1) {
                return helpers.error('string.emailInvalid');
            }

            if (domain.length > UserValidationRule.EMAIL_MAX_DOMAIN_LENGTH) {
                return helpers.error('domainLengthInvalid');
            }

            return value;
        })
        .required()
        .messages({
            'string.email': UserValidationMessage.EMAIL_WRONG,
            'string.empty': UserValidationMessage.EMAIL_REQUIRE,
            'string.emailInvalid': UserValidationMessage.EMAIL_INVALID,
            'domainLengthInvalid': UserValidationMessage.EMAIL_INVALID,
        }),
```

File: shared/src/bundles/users/validation-schemas/email.validation-schema.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import joi from 'joi';
+
+import {
+    type UserForgotPasswordRequestDto,
+    UserValidationMessage,
+} from '../users.js';
+
+const emailValidationSchema = joi.object<UserForgotPasswordRequestDto, true>({
+    email: joi
```
Comment (ValentinaTenus) [inline shared/src/bundles/users/validation-schemas/email.validation-schema.ts:9]:
```suggestion
    email: joi
        .string()
        .email({
            tlds: {
                allow: false,
            },
        })
        .custom((value, helpers) => {
            const localPart = value.split('@')[0];
            const domain = value.split('@')[1];

            if (localPart.length <= 1) {
                return helpers.error('string.emailInvalid');
            }

            if (domain.length > UserValidationRule.EMAIL_MAX_DOMAIN_LENGTH) {
                return helpers.error('domainLengthInvalid');
            }

            return value;
        })
        .required()
        .messages({
            'string.email': UserValidationMessage.EMAIL_WRONG,
            'string.empty': UserValidationMessage.EMAIL_REQUIRE,
            'string.emailInvalid': UserValidationMessage.EMAIL_INVALID,
            'domainLengthInvalid': UserValidationMessage.EMAIL_INVALID,
        }),
```

Comment (ValentinaTenus) [inline shared/src/bundles/users/validation-schemas/email.validation-schema.ts:9]:
```suggestion
   email: joi
        .string()
        .email({
            tlds: {
                allow: false,
            },
        })
        .custom((value, helpers) => {
            const localPart = value.split('@')[0];
            const domain = value.split('@')[1];

            if (localPart.length <= 1) {
                return helpers.error('string.emailInvalid');
            }

            if (domain.length > UserValidationRule.EMAIL_MAX_DOMAIN_LENGTH) {
                return helpers.error('domainLengthInvalid');
            }

            return value;
        })
        .required()
        .messages({
            'string.email': UserValidationMessage.EMAIL_WRONG,
            'string.empty': UserValidationMessage.EMAIL_REQUIRE,
            'string.emailInvalid': UserValidationMessage.EMAIL_INVALID,
            'domainLengthInvalid': UserValidationMessage.EMAIL_INVALID,
        }),
```

File: backend/src/bundles/auth/auth.service.ts [hunk 7]
```diff
@@ -125,6 +133,92 @@ class AuthService implements TAuthService {
...
+                resetPasswordToken,
+                resetPasswordTokenSecret,
+            ) as T;
+        } catch {
+            throw new HTTPError({
+                message: ExceptionMessage.INVALID_RESET_TOKEN,
+                status: HttpCode.BAD_REQUEST,
+            });
+        }
+    }
```
Comment (Icerock10) [inline backend/src/bundles/auth/auth.service.ts:151]: it should be possible to reuse the existing verifyToken method

File: frontend/src/bundles/auth/components/password-recovery/password-recovery.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    const handleEmailSubmit = useCallback(
+        async ({ email }: UserForgotPasswordRequestDto): Promise<void> => {
+            const response = await dispatch(
+                authActions.forgotPassword({ email }),
+            );
+
+            if (response.meta.requestStatus === 'fulfilled') {
+                setSearchParameters({ stage: RecoveryStage.CODE });
+
```
Comment (Icerock10) [inline frontend/src/bundles/auth/components/password-recovery/password-recovery.tsx:35]: use DataStatus enum please

**Comment by ValentinaTenus** [comment]
conflicts

**Review by Icerock10 — CHANGES_REQUESTED** [review]

**Review by Icerock10 — APPROVED** [review]

File: backend/src/bundles/auth/auth.service.ts [hunk 9]
```diff
@@ -125,6 +133,79 @@ class AuthService implements TAuthService {
...
+
+    public async tokenEqualsEmail({
+        email,
+        resetPasswordToken,
+    }: UserVerifyResetPasswordTokenRequestDto): ReturnType<
+        TAuthService['tokenEqualsEmail']
+    > {
+        const user = await this.userService.findByEmail(email);
+
+        const resetPasswordTokenSecret = config.ENV.JWT.RESET_TOKEN_SECRET;
```
Comment (Icerock10) [inline backend/src/bundles/auth/auth.service.ts:145]: You should fill .env.example with the new ENV

**Review by Icerock10 — CHANGES_REQUESTED** [review]

**Review by ValentinaTenus — APPROVED** [review]

**Review by Icerock10 — APPROVED** [review]

**Review by MaxPoplavskiy — APPROVED** [review]

**Review by Vitaliy-Kalenichenko — APPROVED** [review]

**Pull request merged**